### PR TITLE
[CMOB - 2213] Paging

### PR DESCRIPTION
**Pull request by DavitTsiskarashvili**
https://spaceneobank.atlassian.net/browse/CMOB-2213

### CODE REVIEW

File: app/src/main/java/com/movies/data/remote/service/api/ServiceApi.kt [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 interface ServiceApi {
     @GET("{category}")
-    suspend fun getMovies(@Path("category") category: String): Response<MoviesDTO>
+    suspend fun getMovies(@Path("category") category: String, @Query("page") page: Int): Response<MoviesDTO>
 
     @GET(SEARCH)
-    suspend fun searchMovies(@Query("query") query: String): Response<MoviesDTO>
+    suspend fun searchMovies(@Query("query") query: String, @Query("page") page: Int): Response<MoviesDTO>
 
     companion object {
```
Comment (bchmsl) [inline app/src/main/java/com/movies/data/remote/service/api/ServiceApi.kt:16]: Extract endpoints and category paths from each other with prefixes or something

File: app/src/main/java/com/movies/presentation/home/view_model/HomeViewModel.kt [hunk 2]
```diff
@@ -23,53 +28,68 @@ class HomeViewModel(
...
-    private val categoryStateLiveData = MutableLiveData(CategoryType.POPULAR)
+
+    private val _fetchMoviesStateFlow = MutableStateFlow<PagingData<MovieUIModel>?>(null)
+    val fetchMoviesStateFlow get() = _fetchMoviesStateFlow.asStateFlow()
+
+    private val _categoryStateFlow = MutableStateFlow(CategoryType.POPULAR)
+    private val categoryStateFlow = _categoryStateFlow.asStateFlow()
+
+    private val _searchStateFlow = MutableStateFlow<PagingData<MovieUIModel>?>(null)
+    val searchStateFlow = _searchStateFlow.asStateFlow()
```
Comment (bchmsl) [inline app/src/main/java/com/movies/presentation/home/view_model/HomeViewModel.kt:41]: Can you use one StateFlow for fetch movies and searching movies?

File: app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt [hunk 3]
```diff
@@ -28,36 +37,53 @@ class HomeFragment : BaseFragment<HomeViewModel>() {
...
-            movieAdapter.submitList(movies)
-        }
-        observeLiveData(viewModel.searchMoviesLiveData) { searchedMovies ->
-            handleData(searchedMovies.isNotEmpty())
-            movieAdapter.submitList(searchedMovies)
+
+        viewModel.searchStateFlow.collectLatestInLifecycle(viewLifecycleOwner) {
+            handleDataVisibility(true)
+            it.let {
+                if (it != null) {
```
Comment (bchmsl) [inline app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt]: What???

**Review by bchmsl — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt (hunk 3 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt]: done

File: app/src/main/java/com/movies/data/remote/service/api/ServiceApi.kt (hunk 1 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/remote/service/api/ServiceApi.kt:16]: done

File: app/src/main/java/com/movies/presentation/home/view_model/HomeViewModel.kt (hunk 2 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/home/view_model/HomeViewModel.kt:41]: did in network PR

File: app/src/main/java/com/movies/data/remote/paging/MoviesPagingSource.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            search?.let {
+                response = service.searchMovies(it, page)
+            }
+
+            val movies = movieDTOMapper(response.body() ?: throw IllegalArgumentException())
+            val nextKey = if (movies.isEmpty()) {
+                null
+            } else {
+                page + (params.loadSize / 20)
+            }
```
Comment (AnaRusadze) [inline app/src/main/java/com/movies/data/remote/paging/MoviesPagingSource.kt]: i suggest to separate getNextKey into a new function, in order to be more readable

**Review by Gio481 — APPROVED** [review]

File: app/src/main/java/com/movies/data/remote/paging/MoviesPagingSource.kt (hunk 4 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/remote/paging/MoviesPagingSource.kt]: If I separate nextKey into a new function, I won't have access to val page cause it needs params from load function. Therefore, code readability may become even worse. I refactored the existing variable and made it more readable.

**Review by bchmsl — APPROVED** [review]

**Review by etozheluka — APPROVED** [review]

**Review by AnaRusadze — APPROVED** [review]

**Pull request merged**